### PR TITLE
Adding GroupKFold and testing it

### PR DESCRIPTION
**Pull request by hyun-seo**
Referring to #157, there is an issue about group cross-validation.

This pull request implementation this issue by adding new argument `kfolds_group` and
 using `sklearn.model_selection.GroupKFold`.

### CODE REVIEW

File: hpsklearn/estimator/_cost_fn.py [hunk 1]
```diff
@@ -24,6 +25,7 @@ def _cost_fn(argd,
              EX_list: typing.Union[list, tuple] = None,
              valid_size: float = 0.2,
              n_folds: int = None,
+             kfolds_group: typing.Union[list, np.ndarray]  = None,
```
Comment (mandjevant) [inline hpsklearn/estimator/_cost_fn.py:28]:
```suggestion
             kfolds_group: typing.Union[list, np.ndarray] = None,
```

File: hpsklearn/estimator/_cost_fn.py [hunk 2]
```diff
@@ -57,6 +59,11 @@ def _cost_fn(argd,
             n_folds > 2. Or, use leave-one-out cross-validation when
             n_folds = -1.
 
+        kfolds_group: list or ndarray, default is None
+            When kfolds_group is not None, use Group K-fold cross-validation
+            with the specified groups. The length of kfolds_group must be
+            equal to the number of samples in X.
+
```
Comment (mandjevant) [inline hpsklearn/estimator/_cost_fn.py:66]:
```suggestion

```

Comment (mandjevant) [reply hpsklearn/estimator/_cost_fn.py:66]: Leave newline, remove whitespace.

File: hpsklearn/estimator/_cost_fn.py [hunk 3]
```diff
@@ -145,10 +152,18 @@ def _cost_fn(argd,
...
+                    cv_iter = GroupKFold(n_splits=n_folds,
+                                         shuffle=shuffle,
+                                         random_state=random_state_sklearn
+                                         ).split(X, y, kfolds_group)
+                else:
+                    info(f"Will use K-fold CV with K: {n_folds} and Shuffle: {shuffle}")
+                    cv_iter = KFold(n_splits=n_folds,
+                                    shuffle=shuffle,
+                                    random_state=random_state_sklearn).split(X)
+
```
Comment (mandjevant) [inline hpsklearn/estimator/_cost_fn.py:166]:
```suggestion

```

Comment (mandjevant) [reply hpsklearn/estimator/_cost_fn.py:166]: Leave newline, remove whitespace.

File: hpsklearn/estimator/estimator.py [hunk 4]
```diff
@@ -277,6 +283,7 @@ def fit_iter(self, X, y,
                      EX_list=EX_list,
                      valid_size=valid_size,
                      n_folds=n_folds,
+                     kfolds_group =kfolds_group,
```
Comment (mandjevant) [inline hpsklearn/estimator/estimator.py:286]:
```suggestion
                     kfolds_group = kfolds_group,
```

File: hpsklearn/estimator/estimator.py [hunk 5]
```diff
@@ -426,6 +434,11 @@ def fit(self, X, y,
                 n_folds > 2. Or, use leave-one-out cross-validation when
                 n_folds = -1.
 
+            kfolds_group: list or ndarray, default is None
+                When kfolds_group is not None, use Group K-fold cross-validation
+                with the specified groups. The length of group_kfolds must be
+                equal to the number of samples in X.
+
```
Comment (mandjevant) [inline hpsklearn/estimator/estimator.py:441]:
```suggestion

```

Comment (mandjevant) [reply hpsklearn/estimator/estimator.py:441]: Leave newline, remove whitespace.

File: hpsklearn/estimator/estimator.py [hunk 6]
```diff
@@ -242,6 +243,11 @@ def fit_iter(self, X, y,
                 n_folds > 2. Or, use leave-one-out cross-validation when
                 n_folds = -1.
 
+            kfolds_group: list or ndarray, default is None
+                When kfolds_group is not None, use Group K-fold cross-validation
+                with the specified groups. The length of kfolds_group must be
+                equal to the number of samples in X.
+
```
Comment (mandjevant) [inline hpsklearn/estimator/estimator.py:250]:
```suggestion

```

Comment (mandjevant) [reply hpsklearn/estimator/estimator.py:250]: Leave newline, remove whitespace.

File: tests/test_estimator/test_estimator.py [hunk 7]
```diff
@@ -245,6 +245,36 @@ def test_crossvalidation(self):
...
+                np.random.normal(1, 1, size=(1000, 10)),
+            ]),
+            np.random.normal(0, 1, size=(2000, 10)),
+        ])
+        y = np.zeros(2000)
+        y[:1000] = 1
+
+        # Try to fit a model
+        cls = HyperoptEstimator(classifier=sgd_classifier("sgd", loss="log"), preprocessing=[])
+        cls.fit(X, y, cv_shuffle=True, n_folds=5, kfolds_group=[0]*500 + [1]*500 + [2]*500 + [3]*500 + [4]*500)
```
Comment (mandjevant) [inline tests/test_estimator/test_estimator.py:277]:
```suggestion
        cls.fit(X, y, cv_shuffle=True, n_folds=5, kfolds_group=[0]*500 + [1]*500 + [2]*500 + [3]*500 + [4]*500)  # noqa: E226
```

File: tests/test_estimator/test_estimator.py [hunk 8]
```diff
@@ -245,6 +245,36 @@ def test_crossvalidation(self):
...
```
Comment (mandjevant) [inline tests/test_estimator/test_estimator.py:278]:
```suggestion


```

File: tests/test_estimator/test_estimator.py [hunk 9]
```diff
@@ -245,6 +245,36 @@ def test_crossvalidation(self):
         cls = HyperoptEstimator(classifier=sgd_classifier("sgd", loss="log"), preprocessing=[])
         cls.fit(X, y, cv_shuffle=True, n_folds=5)
```
Comment (mandjevant) [inline tests/test_estimator/test_estimator.py:247]:
```suggestion


```

**Review by mandjevant — CHANGES_REQUESTED** [review]
Requesting these initial changes for it to go through the automated testing.
Once it gets through, Ill take a closer look

File: hpsklearn/estimator/estimator.py [hunk 10]
```diff
@@ -277,6 +282,7 @@ def fit_iter(self, X, y,
                      EX_list=EX_list,
                      valid_size=valid_size,
                      n_folds=n_folds,
+                     kfolds_group = kfolds_group,
```
Comment (mandjevant) [inline hpsklearn/estimator/estimator.py:285]:
```suggestion
                     kfolds_group=kfolds_group,
```
My bad. My previous review on this line was incorrect.

Saturday it's on my list to properly look at this pr

File: tests/test_estimator/test_estimator.py [hunk 11]
```diff
@@ -246,5 +246,37 @@ def test_crossvalidation(self):
...
+                np.random.normal(1, 1, size=(1000, 10)),
+            ]),
+            np.random.normal(0, 1, size=(2000, 10)),
+        ])
+        y = np.zeros(2000)
+        y[:1000] = 1
+
+        # Try to fit a model
+        cls = HyperoptEstimator(classifier=sgd_classifier("sgd", loss="log"), preprocessing=[])
+        cls.fit(X, y, cv_shuffle=True, n_folds=5, kfolds_group=np.array([0]*500 + [1]*500 + [2]*500 + [3]*500))  # noqa: E226
```
Comment (mandjevant) [inline tests/test_estimator/test_estimator.py:278]:
```suggestion
        cls.fit(X, y, cv_shuffle=True, n_folds=5,
                kfolds_group=np.array([0]*500 + [1]*500 + [2]*500 + [3]*500))  # noqa: E226
```

**Comment by hyun-seo** [comment]
@mandjevant 
Thank you so much for taking the time to review my pr. 
This is my first attempt at contributing to open source and I had a lot of issues, but I think I learned a lot from your kind help.
I have one question: what is the exact meaning of # noqa: E226 in the testcode? Why is this phrase needed in the test?

**Comment by mandjevant** [comment]
> @mandjevant Thank you so much for taking the time to review my pr. This is my first attempt at contributing to open source and I had a lot of issues, but I think I learned a lot from your kind help. I have one question: what is the exact meaning of # noqa: E226 in the testcode? Why is this phrase needed in the test?

Thank you for your contribution!
The reason your PR did not go through automated testing was because of flake8 style guide enforcement. In most lines, flake was correct and notified us of style issues that we resolved. But in that one line, I deemed that flake8 was wrong. 
It threw an inconsistency with [rule E226](https://www.flake8rules.com/rules/E226.html) which, when enforced, would make the line unreadable.

Thus, I [suppressed the rule](https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html) on (only) that line by adding a comment stating `noqa: E226`.

File: hpsklearn/estimator/estimator.py [hunk 12]
```diff
@@ -426,6 +433,10 @@ def fit(self, X, y,
                 n_folds > 2. Or, use leave-one-out cross-validation when
                 n_folds = -1.
 
+            kfolds_group: list or ndarray, default is None
+                When kfolds_group is not None, use Group K-fold cross-validation
+                with the specified groups. The length of group_kfolds must be
+                equal to the number of samples in X.
```
Comment (mandjevant) [inline hpsklearn/estimator/estimator.py:439]:
```suggestion
                equal to the number of samples in X.

```

File: tests/test_estimator/test_estimator.py [hunk 13]
```diff
@@ -246,5 +246,38 @@ def test_crossvalidation(self):
         cls.fit(X, y, cv_shuffle=True, n_folds=5)
 
 
+class TestGroupCrossValidation(unittest.TestCase):
+    """
+    Class for testing estimator with cross validation
```
Comment (mandjevant) [inline tests/test_estimator/test_estimator.py:251]:
```suggestion
    Class for testing estimator with group cross validation
```

**Review by mandjevant — CHANGES_REQUESTED** [review]
In addition, please change 
```
n_folds: int, default is None
            When n_folds is not None, use K-fold cross-validation when
            n_folds > 2. Or, use leave-one-out cross-validation when
            n_folds = -1.
```
to 

```
n_folds: int, default is None
            When n_folds is not None, use K-fold cross-validation when
            n_folds > 2. Or, use leave-one-out cross-validation when
            n_folds = -1. For Group K-fold cross-validation, functions as
            `n_splits`.
```

In both places:

https://github.com/hyperopt/hyperopt-sklearn/blob/9938225c7d4dd85fbc45be12aa2f74cdde527ad6/hpsklearn/estimator/estimator.py#L244C30-L244C30

https://github.com/hyun-seo/hyperopt-sklearn/blob/9938225c7d4dd85fbc45be12aa2f74cdde527ad6/hpsklearn/estimator/_cost_fn.py#L60C26-L60C26

Function and test case look appropriate

**Review by mandjevant — APPROVED** [review]
Looks good! Thanks for your contribution!
@bjkomer Any thoughts?

**Comment by mandjevant** [comment]
Ill go ahead and merge this. Thank you for your contribution @hyun-seo

**Pull request merged**